### PR TITLE
[feat] Add stronger validation for checkpoint_callback argument

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

This removes the option for passing a `ModelCheckpoint` instance to the checkpoint_callback argument and raises an error if the wrong type is provided. 

This argument is already typed as `bool` in the Trainer, but the callback connector was accepting `Union[bool, ModelCheckpoint]`. This can be very confusing for users who pass model checkpoint callback instances to this argument, as their callback settings are silently ignored. 

This carries forward the deprecation initially landed in https://github.com/PyTorchLightning/pytorch-lightning/pull/4336

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 1]
```diff
@@ -69,7 +70,13 @@ def on_trainer_init(
         # it is important that these are the last callbacks to run
         self.trainer.callbacks = self._reorder_callbacks(self.trainer.callbacks)
 
-    def configure_checkpoint_callbacks(self, checkpoint_callback: Union[ModelCheckpoint, bool]):
+    def _configure_checkpoint_callbacks(self, checkpoint_callback: bool) -> None:
+        if not isinstance(checkpoint_callback, bool):
+            error_msg = "Invalid type provided for checkpoint_callback: "
+            f"Expected bool but received {type(checkpoint_callback)}."
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/callback_connector.py:76]:
```suggestion
            error_msg = (
                "Invalid type provided for checkpoint_callback:"
                f" Expected bool but received {type(checkpoint_callback)}."
            )
```
needs to be surrounded by parentheses, otherwise the second line will not be part of the error_msg

**Review by awaelchli — APPROVED** [review]
nice!

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 2]
```diff
@@ -69,7 +70,13 @@ def on_trainer_init(
         # it is important that these are the last callbacks to run
         self.trainer.callbacks = self._reorder_callbacks(self.trainer.callbacks)
 
-    def configure_checkpoint_callbacks(self, checkpoint_callback: Union[ModelCheckpoint, bool]):
+    def _configure_checkpoint_callbacks(self, checkpoint_callback: bool) -> None:
+        if not isinstance(checkpoint_callback, bool):
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/callback_connector.py:74]: Why do we want to explicitly check this? We don't do it for other properties. Correct typing should be enough for users to see that they are passing the wrong argument type

Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/callback_connector.py:74]: @carmocca purely to alert users since this was previously supported behavior. I can add a TODO to remove this in v1.5 so we rely purely on the type signature. wdyt?

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 3]
```diff
@@ -69,7 +70,15 @@ def on_trainer_init(
...
+            error_msg = (
+                "Invalid type provided for checkpoint_callback:"
+                f" Expected bool but received {type(checkpoint_callback)}."
+            )
+            if isinstance(checkpoint_callback, Callback):
+                error_msg += " Pass callback instances to the `callbacks` argument in the Trainer constructor instead."
+            raise MisconfigurationException(error_msg)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/callback_connector.py:81]:
```suggestion
        # TODO: Remove this error in v1.5 so we rely purely on the type signature
        if not isinstance(checkpoint_callback, bool):
            error_msg = (
                "Invalid type provided for checkpoint_callback:"
                f" Expected bool but received {type(checkpoint_callback)}."
            )
            if isinstance(checkpoint_callback, Callback):
                error_msg += " Pass callback instances to the `callbacks` argument in the Trainer constructor instead."
            raise MisconfigurationException(error_msg)
```

File: tests/checkpointing/test_model_checkpoint.py [hunk 4]
```diff
@@ -1258,3 +1258,9 @@ def test_ckpt_version_after_rerun_same_trainer(tmpdir):
 def test_model_checkpoint_mode_options():
     with pytest.raises(MisconfigurationException, match="`mode` can be .* but got unknown_option"):
         ModelCheckpoint(mode="unknown_option")
+
+
+def test_trainer_checkpoint_callback_bool(tmpdir):
+    mc = ModelCheckpoint(dirpath=tmpdir)
+    with pytest.raises(MisconfigurationException, match="Invalid type provided for checkpoint_callback"):
+        _ = Trainer(checkpoint_callback=mc)
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1266]:
```suggestion
        Trainer(checkpoint_callback=mc)
```

**Review by carmocca — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -22,6 +22,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Changed
 
 
+- Changed the `Trainer`'s `checkpoint_callback` argument to allow boolean values only ([#7539](https://github.com/PyTorchLightning/pytorch-lightning/pull/7539))
```
Comment (ananthsub) [inline CHANGELOG.md:25]:
```suggestion
- Changed the `Trainer`'s `checkpoint_callback` argument to allow only boolean values ([#7539](https://github.com/PyTorchLightning/pytorch-lightning/pull/7539))
```

**Review by tchaton — APPROVED** [review]
LGMT !
